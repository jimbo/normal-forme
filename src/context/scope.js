import { createContext } from "react"

const ScopeContext = createContext("")
const { Consumer, Provider } = ScopeContext

export default ScopeContext
export { Consumer as ScopeConsumer }
export { Provider as ScopeProvider }
