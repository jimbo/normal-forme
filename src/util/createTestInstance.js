import TestRenderer, { act } from "react-test-renderer"

export default (...args) => {
  let instance

  act(() => {
    instance = TestRenderer.create(...args)
  })

  return instance
}
