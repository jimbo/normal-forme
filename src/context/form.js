import { createContext, useCallback, useContext } from "react"

import { getDeepValue } from "../util/deep"

const FormContext = createContext()
const { Consumer, Provider } = FormContext

export default FormContext
export { Consumer as FormConsumer }
export { Provider as FormProvider }

export const useFormContext = field => {
  const [state, dispatch] = useContext(FormContext)
  const value = getDeepValue(state.valueMap, field)

  const setValue = useCallback(
    nextValue => dispatch({ type: "set value", payload: { field, nextValue } }),
    [dispatch, field]
  )

  const transformValue = useCallback(
    transformValue =>
      dispatch({ type: "transform value", payload: { field, transformValue } }),
    [dispatch, field]
  )

  return { setValue, transformValue, value }
}
