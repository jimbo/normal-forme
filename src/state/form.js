import { useReducer } from "react"

import { getDeepValue, setDeepValue } from "../util/deep"

const getInitialState = (initialValueMap = new Map()) => ({
  valueMap: initialValueMap,
})

export const reducer = (state = getInitialState(), { payload, type } = {}) => {
  switch (type) {
    case "set value": {
      const { valueMap } = state
      const { field, nextValue } = payload
      const nextValueMap = setDeepValue(valueMap, field, nextValue)

      return { valueMap: nextValueMap }
    }
    case "transform value": {
      const { valueMap } = state
      const { field, transformValue } = payload
      const currentValue = getDeepValue(valueMap, field)
      const nextValue = transformValue(currentValue)
      const nextValueMap = setDeepValue(valueMap, field, nextValue)

      return { valueMap: nextValueMap }
    }
    default: {
      return state
    }
  }
}

export const useFormState = initialValueMap =>
  useReducer(reducer, getInitialState(initialValueMap))
