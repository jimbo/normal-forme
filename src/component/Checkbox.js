import { createElement, useCallback } from "react"

import { useFormContext } from "../context/form"

const Checkbox = ({ field, value }) => {
  const { transformValue, value: selectedValues } = useFormContext(field)
  const valueSet = selectedValues || new Set()
  const checked = valueSet.has(value)

  const handleChange = useCallback(
    event => {
      const { checked, value } = event.target

      transformValue((valueSet = new Set()) => {
        const nextValueSet = new Set(valueSet)

        if (checked) nextValueSet.add(value)
        else nextValueSet.delete(value)

        return nextValueSet
      })
    },
    [transformValue]
  )

  return (
    <input
      type="checkbox"
      checked={checked}
      value={value}
      onChange={handleChange}
    />
  )
}

export default Checkbox
