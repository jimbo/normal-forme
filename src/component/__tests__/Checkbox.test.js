import { createElement } from "react"

import Checkbox from "../Checkbox"
import { FormProvider } from "../../context/form"
import createTestInstance from "../../util/createTestInstance"

const field = "a"
const dispatch = jest.fn()
const getState = value => ({ valueMap: new Map().set(field, value) })

describe("Checkbox", () => {
  it("renders correctly", () => {
    const state = getState(new Set().add("b"))
    const tree = createTestInstance(
      <FormProvider value={[state, dispatch]}>
        <Checkbox field={field} value="b" />
        <Checkbox field={field} value="c" />
        <Checkbox field={field} value="d" />
      </FormProvider>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it("sets `checked` to `true` if the value matches", () => {
    const selectedValues = new Set().add("c").add("d")
    const state = getState(selectedValues)
    const tree = createTestInstance(
      <FormProvider value={[state, dispatch]}>
        <Checkbox field={field} value="b" />
        <Checkbox field={field} value="c" />
        <Checkbox field={field} value="d" />
      </FormProvider>
    ).toJSON()

    for (const element of tree) {
      const { checked, value } = element.props
      expect(checked).toBe(selectedValues.has(value))
    }
  })

  it("renders if context value is `undefined`", () => {
    const state = getState()
    const tree = createTestInstance(
      <FormProvider value={[state, dispatch]}>
        <Checkbox field={field} value="b" />
        <Checkbox field={field} value="c" />
        <Checkbox field={field} value="d" />
      </FormProvider>
    ).toJSON()

    for (const element of tree) {
      expect(element.props.checked).toBe(false)
    }
  })

  it("renders if context value is `null`", () => {
    const state = getState(null)
    const tree = createTestInstance(
      <FormProvider value={[state, dispatch]}>
        <Checkbox field={field} value="b" />
        <Checkbox field={field} value="c" />
        <Checkbox field={field} value="d" />
      </FormProvider>
    ).toJSON()

    for (const element of tree) {
      expect(element.props.checked).toBe(false)
    }
  })

  it("dispatches on change", () => {
    const valueToAppend = "d"
    const state = getState(new Set().add("b"))
    const tree = createTestInstance(
      <FormProvider value={[state, dispatch]}>
        <Checkbox field={field} value="b" />
        <Checkbox field={field} value="c" />
        <Checkbox field={field} value="d" />
      </FormProvider>
    ).toJSON()

    const nextElement = tree.find(({ props }) => props.value === valueToAppend)
    nextElement.props.onChange({
      target: { checked: true, value: valueToAppend },
    })

    expect(dispatch).toHaveBeenCalledTimes(1)
    expect(dispatch).toHaveBeenLastCalledWith({
      payload: { field, transformValue: expect.any(Function) },
      type: "transform value",
    })
  })
})

describe("Checkbox's transformValue", () => {
  it("includes a checked value in the value set", () => {
    const valueToAppend = "d"
    const state = getState(new Set().add("b"))
    const tree = createTestInstance(
      <FormProvider value={[state, dispatch]}>
        <Checkbox field={field} value="b" />
        <Checkbox field={field} value="c" />
        <Checkbox field={field} value="d" />
      </FormProvider>
    ).toJSON()

    const nextElement = tree.find(({ props }) => props.value === valueToAppend)
    nextElement.props.onChange({
      target: { checked: true, value: valueToAppend },
    })

    const { transformValue } = dispatch.mock.calls[0][0].payload
    const valueSet = state.valueMap.get(field)
    const result = transformValue(valueSet)

    expect(result).toEqual(new Set(valueSet).add(valueToAppend))
  })

  it("removes an unchecked value from the value set", () => {
    const valueToRemove = "d"
    const state = getState(new Set().add("c").add("d"))
    const tree = createTestInstance(
      <FormProvider value={[state, dispatch]}>
        <Checkbox field={field} value="b" />
        <Checkbox field={field} value="c" />
        <Checkbox field={field} value="d" />
      </FormProvider>
    ).toJSON()

    const nextElement = tree.find(({ props }) => props.value === valueToRemove)
    nextElement.props.onChange({
      target: { checked: false, value: valueToRemove },
    })

    const { transformValue } = dispatch.mock.calls[0][0].payload
    const valueSet = state.valueMap.get(field)
    const result = transformValue(valueSet)
    const expected = new Set(valueSet)

    expected.delete(valueToRemove)
    expect(result).toEqual(expected)
  })

  it("creates a new value set if it's undefined", () => {
    const valueToAppend = "d"
    const state = getState()
    const tree = createTestInstance(
      <FormProvider value={[state, dispatch]}>
        <Checkbox field={field} value="b" />
        <Checkbox field={field} value="c" />
        <Checkbox field={field} value="d" />
      </FormProvider>
    ).toJSON()

    const nextElement = tree.find(({ props }) => props.value === valueToAppend)
    nextElement.props.onChange({
      target: { checked: true, value: valueToAppend },
    })

    const { transformValue } = dispatch.mock.calls[0][0].payload
    const result = transformValue()

    expect(result).toEqual(new Set().add(valueToAppend))
  })
})
