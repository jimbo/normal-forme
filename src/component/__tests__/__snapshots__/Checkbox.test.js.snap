// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
Array [
  <input
    checked={true}
    onChange={[Function]}
    type="checkbox"
    value="b"
  />,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
    value="c"
  />,
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
    value="d"
  />,
]
`;
